src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/transformer/CalendarTransformer.kt

```diff
@@ -0,0 +1,73 @@
+package uk.gov.justice.digital.hmpps.approvedpremisesapi.transformer
+
+import org.springframework.stereotype.Component
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.api.model.BedOccupancyBookingEntry
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.api.model.BedOccupancyEntryType
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.api.model.BedOccupancyLostBedEntry
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.api.model.BedOccupancyOpenEntry
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.api.model.BedOccupancyRange
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.repository.CalendarBedInfo
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.repository.CalendarBookingInfo
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.repository.CalendarLostBedInfo
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.repository.CalendarOccupancyInfo
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.util.getDaysUntilInclusive
+import java.time.LocalDate
+
+@Component
+class CalendarTransformer {
+  fun transformDomainToApi(startDate: LocalDate, endDate: LocalDate, domain: Map<CalendarBedInfo, List<CalendarOccupancyInfo>>): List<BedOccupancyRange> {
+    return domain.entries.map { bedCalendarInfo ->
+      val transformedSchedule = bedCalendarInfo.value.map {
+        when (it) {
+          is CalendarBookingInfo -> BedOccupancyBookingEntry(
+            bookingId = it.bookingId,
+            personName = it.personName!!,
+            type = BedOccupancyEntryType.booking,
+            length = it.startDate.getDaysUntilInclusive(it.endDate).size,
+            startDate = it.startDate,
+            endDate = it.endDate,
+          )
+          is CalendarLostBedInfo -> BedOccupancyLostBedEntry(
+            lostBedId = it.lostBedId,
+            type = BedOccupancyEntryType.lostBed,
+            length = it.startDate.getDaysUntilInclusive(it.endDate).size,
+            startDate = it.startDate,
+            endDate = it.endDate,
+          )
+        }
+      }.toMutableList()
+
+      var startOfOpenPeriod: LocalDate? = null
+      var endOfOpenPeriod: LocalDate? = null
+      startDate.getDaysUntilInclusive(endDate).forEach { dateInRange ->
```

Comment:
I'm finding this block of code a little hard to follow, I'm wondering if we can do it in a different way by getting all the dates for the booked periods (Not sure if this is valid Kotlin!):

```kotlin
val bookedDates = transformedSchedule.map { it.startDate.getDaysUntilInclusive(it.endDate) }
val entirePeriod = startDate.getDaysUntilInclusive(endDate)

// Get all the missing dates in the schedule
val openDates = entirePeriod.minus(bookedDates)

openDates.forEach { i, startDate -> 
  val endDate: LocalDate? = null

  // Loop through all the dates following the current date
  openDates.drop(i + 1).forEach { length, date -> 
    // If the date is not the next day, we set next day to the previous day in the loop
    if (endDate !== startDate.addDays(length + 1)) {
      endDate = openDates[i -1]
    }
  }

 transformedSchedule += BedOccupancyOpenEntry(
   type = BedOccupancyEntryType.open,
   length = startDate!!.getDaysUntilInclusive(endDate!!).size,
   startDate = startDate!!,
   endDate = endDate!!,
 )
}
```

Reply:
Wouldn't this add multiple entries for the same open period with the start date moving up by a day for each entry?